### ISSUE TITLE
Feature Request: `--unbuffered` flag for the Chapel Slurm launcher

### ISSUE THREAD

**Issue by kwaters4** [chapel-lang/chapel#22681]
### Summary of Problem
When writing and debugging an HPC application I found that I quickly ran away from the Chapel Slurm launcher. The main reason was the inability to pass --unbuffered to Slurm from the Chapel Slurm launcher. 

**Work Around**: 
Use the Chapel Slurm launcher once with the `-v` argument and use the output (using `foo_real` instead of `foo`) plus the `--unbuffered` flag to launch Chapel programs.

**Best Solution(?)**:
Have a flag that can be used with the Chapel Launcher that will pass arguments directly to Slurm?

This is a small quality of life fix, nothing is blocking here.

**Comment by bradcray** [comment]
@jhh67 and @ronawho : I can never remember: Do we have a general `CHPL_LAUNCHER_OPTS`-style variable that permits a user to specify additional arbitrary flags to pass through to the underlying launcher (which could be another way of providing this capability)?  Taking a quick scan through the sources and docs, I'm thinking 'no', but it was admittedly fast.

**Comment by ronawho** [comment]
No, but I think that would be worth implementing.

In the meantime I tend to just use the slurm environment variables when they're available. In this case it should be `SLURM_UNBUFFEREDIO=1`

**Comment by bradcray** [comment]
A question that came up in discussing this a bit with @kwaters4 is whether it would be more productive for users to have unbuffered be the default for slurm launches from a productivity standpoint.  The case he ran into was one in which he'd done a bunch of debug printing within a timestep loop, say ("on iteration 1", "on iteration 2", etc.), then hit a compute-intensive portion of the code, and the buffered mode caused it to appear as though his computation got hung up on the n-k'th iteration, when in fact it had completed the timestep loop, but the slurm buffering had prevented it from printing.

The obvious downside to this would be the potential of performance impact, but maybe it would still make sense to have the productive behavior first and performance second.  I suggested we could tie buffered/unbuffered to whether or not `--fast` was used, but he explained that he almost always uses fast, so would really want the ability to say fast + unbuffered in any case (whatever the defaults were).

**Comment by ronawho** [comment]
Unbuffered I/O can have a pretty large and negative impact on the WLM if there's a lot of output, so I'd be nervous about making that the default.

I'd be curious if just adding a flush to the debug prints made them appear at the "right time". Assuming stdout, I think that's:

```chpl
try! stdout.flush();
```

**Comment by bradcray** [comment]
I was thinking that Chapel (which is to say, the C code that Chapel relies upon) would be "doing the right thing" w.r.t. buffering and flushing and that the slurm option was enabling/disabling some sort of additional buffering done on top of that.  But Elliot, it sounds like you're saying they're one and the same (e.g., maybe slurm is turning off buffering at the C library level?  I tried to read the docs for the option, but didn't come out with much confidence).

@kwaters4 : If you get a chance to try Elliot's experiment (with buffering in its default/enabled state), I'd be curious what you find as well.

**Comment by ronawho** [comment]
> But Elliot, it sounds like you're saying they're one and the same

No, I was just wondering if flushing would do anything. I'm not really familiar with buffering or IO enough to say anything confidently. My vague understanding is that chapel/c do line buffering by default and slurm does some additional buffering on top of that, but I could still imagine flushing would go through both.

Testing on an internal slurm system flushing does seem to help:

```chpl
use Time, IO;

config const trials = 3;
config const flushIt = false;
config const sleepTime = 3.0;

var i: int;
proc writeIt() {
  writeln(i); i+=1;
  if flushIt then try! stdout.flush();
}

for i in 1..trials {
  writeIt();
  sleep(sleepTime);
}
```

I see output in "real time" with either `SLURM_UNBUFFEREDIO=1` or with `--flushIt` enabling `stdout.flush()`. Without either if I bump the number of trials enough I eventually see output. Testing out with larger chunks of output, it's probably 4K bytes that triggers a write, so I'd guess something is page buffering.

**Comment by kwaters4** [comment]
@ronawho My understanding here is the same. My intuition tells me that Slurm and stdout are working on two separate buffers here. I think the two will diverge if you run across multiple nodes. 

Slurm Manual:

> By default, the connection between slurmstepd and the user-launched application is over a pipe. The stdio output written by the application is buffered by the glibc until it is flushed or the output is set as unbuffered. See [setbuf](https://slurm.schedmd.com/setbuf.html)(3). If this option is specified the tasks are executed with a pseudo terminal so that the application output is unbuffered. This option applies to step allocations.

Going into the weeds here and a tangent here.
1) Would all the print statements first have to be communicated to Locale 0 and then placed into Locale 0's local buffer?
or
2) Is it first buffered on Locale N-1 and then once that buffer is full it migrates to Locale 0. 

In terms of Chapel is `stdout.flush()` the same is `--unbuffered` for Case 1? There is a nuance here I think, but I am unfamiliar with the stdout plumbing of Slurm. Doing some test with multiple nodes should(?) answer the question. I may get to it next week.

**Comment by bradcray** [comment]
With respect to the tangents:

Today, I believe all write*() routines on stdout involve proactive communication back to locale 0 to get the output generated in a reasonably ordered, atomic, and timely manner.  There've been musings at times about making the IO lazier / locally buffered on a locale / etc., but nothing along those lines has been implemented that I'm aware of (and I think I would be).  For that reason, I'd expect any buffering to be strictly about how locale 0 printed things through C, what bufering C did, and then any additional buffering that slurm (or whatever) did on top of C.

For that reason, I'd expect that `--unbuffered` isn't necessarily the same as `stdout.flush()` since the former may have slurm-specific buffering involved and the latter is unaware of slurm.

I wouldn't expect differences in how writeln() and writef() do buffering at the Chapel level, though I could imagine potential behavior differences if we're relying on (or affected by) line-based buffering (in C, e.g.) and a writef() were to not contain / end with a `"\n"` character.

Caveat: All of these statements are impressions though, and could be wrong or outdated...  And I'm definitely willing to look into them more (or ask someone else to) if you're seeing things that suggest that something doesn't seem right.

**Comment by bradcray** [comment]
Just noting here that #28380 (not yet merged) has implemented this.

**Comment by bradcray** [comment]
@jabraham17 : I think you meant for this to be closed by the merge of #28380, but that GitHub's parser didn't handle the "…and" such that it wasn't?  If so, might be worth checking whether other PRs you've merged only resulted in similar partial closings.

**Comment by jabraham17** [comment]
Yes, https://github.com/chapel-lang/chapel/pull/28380 should have closed this.

I usually write `resolves #...` separately for this reason. I guess I forgot this time
